Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+HTTP2Connections.swift

```diff
@@ -0,0 +1,446 @@
+        /// A request can be scheduled on the connection
+        var isAvailable: Bool {
+            switch self.state {
+            case .active(_, let maxStreams, let usedStreams, _):
+                return usedStreams < maxStreams
+            case .starting, .backingOff, .draining, .closed:
+                return false
+            }
+        }
+
+        /// The connection is active but there are no requests running on the connection
+        var isIdle: Bool {
+            switch self.state {
+            case .active(_, _, let usedStreams, _):
+                return usedStreams == 0
+            case .starting, .backingOff, .draining, .closed:
+                return false
+            }
+        }
+        
+        var isClosed: Bool {
+            switch self.state {
+            case .starting, .backingOff, .draining, .active:
+                return false
+            case .closed:
+                return true
+            }
+        }
+
+        var usedAndMaxStreams: (Int, Int)? {
+            switch self.state {
+            case .starting, .backingOff, .draining, .closed:
+                return nil
+            case .active(_, let maxStreams, let usedStreams, _):
+                return (usedStreams, maxStreams)
+            }
+        }
+        
+        var connection: Connection? {
+            switch state {
+            case .starting, .backingOff, .closed:
+                return nil
+            case .active(let connection, _, _, _),
+                    .draining(let connection, _, _):
+                return connection
+            }
+        }
+
+        private var state: State
+        let eventLoop: EventLoop
+        let connectionID: Connection.ID
+
+        /// should be called after the connection was successfully established
+        /// - Parameters:
+        ///   - conn: HTTP2 connection
+        ///   - maxStreams: max streams settings from the server
+        /// - Returns: number of available streams which can be leased
+        mutating func connected(_ conn: Connection, maxStreams: Int) -> Int {
+            guard case .starting = self.state else {
+                preconditionFailure("Invalid state")
+            }
+            self.state = .active(conn, maxStreams: maxStreams, usedStreams: 0, lastIdle: .now())
+            return maxStreams
+        }
+        
+        /// should be called after receiving new http2 settings from the server
+        /// - Parameters:
+        ///   - maxStreams: max streams settings from the server
+        /// - Returns: number of available streams which can be leased
+        mutating func newMaxConcurrentStreams(_ maxStreams: Int) -> Int {
+            guard case let .active(conn, _, usedStreams, lastIdle) = self.state else {
+                preconditionFailure("Invalid state")
+            }
+            self.state = .active(conn, maxStreams: maxStreams, usedStreams: usedStreams, lastIdle: lastIdle)
+            return max(maxStreams - usedStreams, 0)
```

Comment:
I think, in the draining case, we are probably in a position where we could receive further setting updates.

```swift
            switch self.state {
            case let .active(conn, _, usedStreams, lastIdle):
                self.state = .active(conn, maxStreams: maxStreams, usedStreams: usedStreams, lastIdle: lastIdle)
                return max(maxStreams - usedStreams, 0)
            case others:
                preconditionFailure("Invalid state: \(self.state)")
            }
```

however if we are draining the available streams should always be 0.